### PR TITLE
NDEV-2059: Switch to default ZSTD(1) for every table, add rooted_slots table

### PR DESCRIPTION
**Pull request by Deniskore**
In order to reduce CPU usage on ClickHouse servers and improve SQL execution speed, we need to switch back to LZ4 compression (default compressor)

Consequences:
+ Better decompression speed 
+ Faster SQL execution speed and correspondingly lower CPU usage.
- Higher disk space usage.

Benchmarks from Facebook
https://facebook.github.io/zstd/#benchmarks

We will test it on our testing environment.

### CODE REVIEW

**Comment by Deniskore** [comment]
I ran a local benchmark on a 300MB file (randomly generated data) using default compression levels:

```
Compressing with LZ4...
real	0m0.332s
user	0m0.040s
sys	0m0.291s

Compressing with Zstandard...
real	0m0.662s
user	0m0.265s
sys	0m0.297s
Compression completed.
```

```
Decompressing with LZ4...
real	0m0.378s
user	0m0.120s
sys	0m0.216s

Decompressing with Zstandard...
real	0m0.594s
user	0m0.068s
sys	0m0.243s
Decompression completed.
```

Clickhouse uses custom implementation of streaming for LZ4 and ZSTD.
While results may vary, LZ4 is certainly the winner in terms of compression\decompression speed.

File: v2-advanced/clickhouse/create_rooted_slots.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    parent Nullable(UInt64) default 0 CODEC(DoubleDelta, LZ4),
+) ENGINE = ReplicatedReplacingMergeTree(
+    '/clickhouse/tables/rooted_slots',
+    '{replica}'
+) PRIMARY KEY (slot)
+PARTITION BY toInt32(slot / 216000)
+ORDER BY (slot)
+SETTINGS index_granularity=8192;
+
+CREATE MATERIALIZED VIEW IF NOT EXISTS events.rooted_slots_mv ON CLUSTER '{cluster}' TO events.rooted_slots AS
```
Comment (dk-neon) [inline v2-advanced/clickhouse/create_rooted_slots.sql:14]: Do we really need this materialized view if we already have a separated table for rooted slots?

**Comment by Deniskore** [comment]
After conducting further research, it appears that ZSTD(1) is the most suitable option for our needs.

On local machine file with randomly generated data.
```
Compressing with Zstandard level 1...

real    0m0.351s
user    0m0.191s
sys     0m0.378s
314580014 zstd_compressed.bin


Compressing with LZ4 level 1...
Compressed 314572800 bytes into 314573115 bytes ==> 100.00%                    

real    0m0.970s
user    0m0.007s
sys     0m0.375s
314573115 lz4_compressed.bin
Compression completed.
```

File with 300mb of zeros.
```
Compressing with Zstandard level 1...

real    0m0.098s
user    0m0.073s
sys     0m0.108s
10666 zstd_compressed.bin


Compressing with LZ4 level 1...
Compressed 314572800 bytes into 1234740 bytes ==> 0.39%                        

real    0m0.127s
user    0m0.038s
sys     0m0.088s
1234740 lz4_compressed.bin
Compression completed.
```

I triggered a full scan for update_slot table with different compressor type:

ZSTD(1)
```
production node 1 :) SELECT 1 IN (
        SELECT slot
        FROM events.update_slot_zstd1
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

SELECT 1 IN (
        SELECT slot
        FROM events.update_slot_zstd1
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

Query id: 4c63052b-4dc9-43f7-97ae-e76ddc875c6c

┌─in(1, _subquery1918662)─┐
│                       0 │
└─────────────────────────┘

1 row in set. Elapsed: 3.956 sec. Processed 47.08 million rows, 410.84 MB (11.90 million rows/s., 103.85 MB/s.)
```

ZSTD(5)
```
production node 1 :) SELECT 1 IN (
        SELECT slot
        FROM events.update_slot
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

SELECT 1 IN (
        SELECT slot
        FROM events.update_slot
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

Query id: 98260138-7e33-432b-b1f5-77b84d569353

┌─in(1, _subquery1918663)─┐
│                       0 │
└─────────────────────────┘

1 row in set. Elapsed: 4.008 sec. Processed 47.08 million rows, 410.85 MB (11.75 million rows/s., 102.50 MB/s.)
```
LZ4(default)
```
production node 1 :) SELECT 1 IN (
        SELECT slot
        FROM events.update_slot_lz4
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

SELECT 1 IN (
        SELECT slot
        FROM events.update_slot_lz4
        WHERE status = 'Rooted'
        ORDER BY slot DESC
    )

Query id: f3e55bf3-c098-4a93-8456-8b38b877b908

┌─in(1, _subquery1918664)─┐
│                       0 │
└─────────────────────────┘

1 row in set. Elapsed: 4.158 sec. Processed 45.86 million rows, 409.57 MB (11.03 million rows/s., 98.50 MB/s.)
```

The results are pretty stable, ZSTD(1) always a winner.
I want to hear from @stanislav-tkach about his experiments on local dev env.

File: v2-advanced/clickhouse/create_rooted_slots.sql (hunk 1 above)
Comment (Deniskore) [reply v2-advanced/clickhouse/create_rooted_slots.sql:14]: Yes, we want to update this table as soon as the new data is inserted into the update_slot. We need a separate MV for this purpose.
If there are other ideas, I'd listen, but this implementation is the standard the ClickHouse developers recommend.

**Comment by andreisilviudragnea** [comment]
@Deniskore I think the commit with `rooted_slots` table should be merged in a separate PR, since the changes from https://github.com/neonlabsorg/neon-evm/pull/181 already got merged.

And the new table needs to be created on Mainnet Clickhouse servers, in order to avoid problems with neon-api not finding the table in the database when the new version will be deployed.

**Comment by stanislav-tkach** [comment]
Here are the measurement results:

#### CODEC(DoubleDelta, LZ4)

1 row in set. Elapsed: 1.087 sec. Processed 37.46 million rows, 337.13 MB (34.46 million rows/s., 310.14 MB/s.)
1 row in set. Elapsed: 0.773 sec. Processed 37.46 million rows, 337.13 MB (48.48 million rows/s., 436.29 MB/s.)
1 row in set. Elapsed: 0.773 sec. Processed 37.46 million rows, 337.13 MB (48.49 million rows/s., 436.41 MB/s.)
1 row in set. Elapsed: 0.769 sec. Processed 37.46 million rows, 337.13 MB (48.69 million rows/s., 438.17 MB/s.)
1 row in set. Elapsed: 0.751 sec. Processed 37.46 million rows, 337.13 MB (49.89 million rows/s., 449.01 MB/s.)
1 row in set. Elapsed: 0.790 sec. Processed 37.46 million rows, 337.13 MB (47.43 million rows/s., 426.87 MB/s.)
1 row in set. Elapsed: 0.830 sec. Processed 37.46 million rows, 337.13 MB (45.11 million rows/s., 405.95 MB/s.)
1 row in set. Elapsed: 0.737 sec. Processed 37.46 million rows, 337.13 MB (50.81 million rows/s., 457.25 MB/s.)
1 row in set. Elapsed: 0.740 sec. Processed 37.46 million rows, 337.13 MB (50.62 million rows/s., 455.60 MB/s.)
1 row in set. Elapsed: 1.058 sec. Processed 37.46 million rows, 337.13 MB (35.39 million rows/s., 318.50 MB/s.)

#### CODEC(DoubleDelta, ZSTD(5))

1 row in set. Elapsed: 0.929 sec. Processed 40.74 million rows, 366.62 MB (43.87 million rows/s., 394.85 MB/s.)
1 row in set. Elapsed: 0.769 sec. Processed 40.74 million rows, 366.62 MB (52.97 million rows/s., 476.72 MB/s.)
1 row in set. Elapsed: 0.770 sec. Processed 40.74 million rows, 366.62 MB (52.89 million rows/s., 476.00 MB/s.)
1 row in set. Elapsed: 0.795 sec. Processed 40.74 million rows, 366.62 MB (51.27 million rows/s., 461.45 MB/s.)
1 row in set. Elapsed: 0.827 sec. Processed 40.74 million rows, 366.62 MB (49.25 million rows/s., 443.29 MB/s.)
1 row in set. Elapsed: 0.749 sec. Processed 40.74 million rows, 366.62 MB (54.38 million rows/s., 489.44 MB/s.)
1 row in set. Elapsed: 0.795 sec. Processed 40.74 million rows, 366.62 MB (51.22 million rows/s., 460.99 MB/s.)
1 row in set. Elapsed: 0.751 sec. Processed 40.74 million rows, 366.62 MB (54.28 million rows/s., 488.50 MB/s.)
1 row in set. Elapsed: 0.752 sec. Processed 40.74 million rows, 366.62 MB (54.14 million rows/s., 487.29 MB/s.)
1 row in set. Elapsed: 0.867 sec. Processed 40.74 million rows, 366.62 MB (46.97 million rows/s., 422.70 MB/s.)

#### CODEC(DoubleDelta, ZSTD(1))

1 row in set. Elapsed: 1.194 sec. Processed 42.80 million rows, 385.18 MB (35.84 million rows/s., 322.60 MB/s.)
1 row in set. Elapsed: 0.831 sec. Processed 42.80 million rows, 385.18 MB (51.51 million rows/s., 463.56 MB/s.)
1 row in set. Elapsed: 0.836 sec. Processed 42.80 million rows, 385.18 MB (51.17 million rows/s., 460.57 MB/s.)
1 row in set. Elapsed: 0.893 sec. Processed 42.80 million rows, 385.18 MB (47.93 million rows/s., 431.37 MB/s.)
1 row in set. Elapsed: 0.820 sec. Processed 42.80 million rows, 385.18 MB (52.19 million rows/s., 469.67 MB/s.)
1 row in set. Elapsed: 0.915 sec. Processed 42.80 million rows, 385.18 MB (46.76 million rows/s., 420.80 MB/s.)
1 row in set. Elapsed: 1.032 sec. Processed 42.80 million rows, 385.18 MB (41.45 million rows/s., 373.09 MB/s.)
1 row in set. Elapsed: 0.837 sec. Processed 42.80 million rows, 385.18 MB (51.15 million rows/s., 460.35 MB/s.)
1 row in set. Elapsed: 0.831 sec. Processed 42.80 million rows, 385.18 MB (51.50 million rows/s., 463.51 MB/s.)
1 row in set. Elapsed: 0.817 sec. Processed 42.80 million rows, 385.18 MB (52.37 million rows/s., 471.36 MB/s.)

#### CODEC(ZSTD(1))

1 row in set. Elapsed: 0.467 sec. Processed 37.99 million rows, 341.90 MB (81.32 million rows/s., 731.89 MB/s.)
1 row in set. Elapsed: 0.521 sec. Processed 37.99 million rows, 341.90 MB (72.90 million rows/s., 656.09 MB/s.)
1 row in set. Elapsed: 0.494 sec. Processed 37.99 million rows, 341.90 MB (76.92 million rows/s., 692.31 MB/s.)
1 row in set. Elapsed: 0.511 sec. Processed 37.99 million rows, 341.90 MB (74.31 million rows/s., 668.79 MB/s.)
1 row in set. Elapsed: 0.449 sec. Processed 37.99 million rows, 341.90 MB (84.64 million rows/s., 761.77 MB/s.)
1 row in set. Elapsed: 0.515 sec. Processed 37.99 million rows, 341.90 MB (73.83 million rows/s., 664.48 MB/s.)
1 row in set. Elapsed: 0.487 sec. Processed 37.99 million rows, 341.90 MB (78.06 million rows/s., 702.55 MB/s.)
1 row in set. Elapsed: 0.516 sec. Processed 37.99 million rows, 341.90 MB (73.58 million rows/s., 662.26 MB/s.)
1 row in set. Elapsed: 0.504 sec. Processed 37.99 million rows, 341.90 MB (75.39 million rows/s., 678.47 MB/s.)
1 row in set. Elapsed: 0.494 sec. Processed 37.99 million rows, 341.90 MB (76.90 million rows/s., 692.08 MB/s.)

**Comment by Deniskore** [comment]
Based on million rows/s and total throughput, we are switching to CODEC(ZSTD(1))
Objections?
@dk-neon 
@andreisilviudragnea
@stanislav-tkach

**Comment by stanislav-tkach** [comment]
I think that performance improvements are very obvious for `CODEC(ZSTD(1))`, so I am all for switching to this variant.

**Comment by andreisilviudragnea** [comment]
I am ok with the change. But is changing the compression type easy to do on an existing database with existing data?

**Comment by Deniskore** [comment]
> I am ok with the change. But is changing the compression type easy to do on an existing database with existing data?

Unfortunately no, we will smoothly apply the changes first to the test environment, then to Devnet, and then we think about what to do on Mainnet.

**Review by andreisilviudragnea — APPROVED** [review]

**Comment by Deniskore** [comment]
Looks like GitHub won't let you rename the branch and leave the pull request open...

**Review by stanislav-tkach — APPROVED** [review]

**Pull request merged**